parlai/agents/transformer/modules.py

```diff
@@ -295,34 +341,55 @@ def __init__(
                 attention_dropout=attention_dropout,
                 relu_dropout=relu_dropout,
                 dropout=dropout,
+                variant=variant,
+                activation=activation,
             ))
 
-    def forward(self, input):
+    def forward(self, input, positions=None, segments=None):
         """
             input data is a FloatTensor of shape [batch, seq_len, dim]
             mask is a ByteTensor of shape [batch, seq_len], filled with 1 when
             inside the sequence and 0 outside.
         """
         mask = input != self.padding_idx
-        positions = (mask.cumsum(dim=1, dtype=torch.int64) - 1).clamp_(min=0)
+        if positions is None:
+            positions = (mask.cumsum(dim=1, dtype=torch.int64) - 1).clamp_(min=0)
         tensor = self.embeddings(input)
         if self.embeddings_scale:
             tensor = tensor * np.sqrt(self.dim)
+
         tensor = tensor + self.position_embeddings(positions).expand_as(tensor)
+
+        if self.n_segments >= 1:
+            if segments is None:
+                segments = torch.zeros_like(input)
+            tensor = tensor + self.segment_embeddings(segments)
+
+        if self.variant == 'xlm':
+            tensor = _normalize(tensor, self.norm_embeddings)
+
         # --dropout on the embeddings
         tensor = self.dropout(tensor)
 
         tensor *= mask.unsqueeze(-1).type_as(tensor)
         for i in range(self.n_layers):
             tensor = self.layers[i](tensor, mask)
 
-        if self.reduction:
-            divisor = mask.type_as(tensor).sum(dim=1).unsqueeze(-1).clamp(min=1e-7)
+        if self.reduction_type == 'first':
+            return tensor[:, 0, :]
+        elif self.reduction_type == 'max':
+            return tensor.max(dim=1)[0]
+        elif self.reduction_type == 'mean':
+            divisor = mask.float().sum(dim=1).unsqueeze(-1).clamp(min=1e-20)
             output = tensor.sum(dim=1) / divisor
             return output
-        else:
+        elif self.reduction_type == 'none' or self.reduction_type is None:
```

Comment:
'none' is kinda weird, what is this here for? is this legacy stuff?

Reply:
So the main None is actually what the generative model needs to ensure there's no reduction. The `'none'` was just imagining that it *could* be passed as something... eventually.